### PR TITLE
[InstCombine] Fix a cycle when folding fneg(select) with scalable vector types

### PR DESCRIPTION
**Pull request by ssijaric-nv**
The two folding operations are causing a cycle for the following case with
scalable vector types:

define <vscale x 2 x double> @test_fneg_select_abs(<vscale x 2 x i1> %cond, <vscale x 2 x double> %b) {
  %1 = select <vscale x 2 x i1> %cond, <vscale x 2 x double> zeroinitializer, <vscale x 2 x double> %b
  %2 = fneg fast <vscale x 2 x double> %1
  ret <vscale x 2 x double> %2
}

1) fold fneg:  -(Cond ? C : Y) -> Cond ? -C : -Y

2) fold select: (Cond ? -X : -Y) -> -(Cond ? X : Y)

1) results in the following since '<vscale x 2 x double> zeroinitializer' passes
the check for the immediate constant:

%.neg = fneg fast <vscale x 2 x double> zeroinitializer
%b.neg = fneg fast <vscale x 2 x double> %b
%1 = select fast <vscale x 2 x i1> %cond, <vscale x 2 x double> %.neg, <vscale x 2 x double> %b.neg

and so we end up going back and forth between 1) and 2).

Attempt to fold scalable vector constants, so that we end up with a splat instead:

define <vscale x 2 x double> @test_fneg_select_abs(<vscale x 2 x i1> %cond, <vscale x 2 x double> %b) {
  %b.neg = fneg fast <vscale x 2 x double> %b
  %1 = select fast <vscale x 2 x i1> %cond, <vscale x 2 x double> shufflevector (<vscale x 2 x double> insertelement (<vscale x 2 x double> poison, double -0.000000e+00, i64 0), <vscale x 2 x double> poison, <vscale x 2 x i32> zeroinitializer), <vscale x 2 x double> %b.neg
  ret <vscale x 2 x double> %1
}

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/fneg.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fneg.ll:1123]: Test with zeroinitializer in both positions

**Comment by arsenm** [comment]
> 1) if we are dealing with scalable vector constants as there is no
immediate representation for a negative scalable zeroinitializer.

I thought there was a splat constant for these, this would just be a splat of -0

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2881]: We can use `ConstantFoldUnaryOpOperand(Instruction::FNeg, C, DL)` instead of using IRBuilder to fix this issue.

if (match(X, m_ImmConstant(XC)) || match(Y, m_ImmConstant(YC))) {
   Value *NegX = ConstantFoldUnaryOpOperand(Instruction::FNeg, XC, DL);
   Value *NegY = ConstantFoldUnaryOpOperand(Instruction::FNeg, YC, DL);
   if (NegX || NegY) {
      if (!NegX) NegX = Builder.CreateFNeg...;
      if (!NegY) NegY = Builder.CreateFNeg...;
      SelectInst *NewSel = SelectInst::Create(Cond, NegX, NegY);
      propagateSelectFMF(NewSel, /*CommonOperand=*/true);
      return NewSel;
   }
}

**Comment by ssijaric-nv** [comment]
> > 1. if we are dealing with scalable vector constants as there is no
> >    immediate representation for a negative scalable zeroinitializer.
> 
> I thought there was a splat constant for these, this would just be a splat of -0

Thanks for having a look, Matt.  Would a better approach be to generate the following, instead of giving up?

define <vscale x 2 x double> @test_fneg_select_svec(<vscale x 2 x i1> %cond, <vscale x 2 x double> %b) {
  %1 = fneg  fast <vscale x 2 x double> %b
  %2 = select <vscale x 2 x i1> %cond, <vscale x 2 x double> splat(double -0.0), <vscale x 2 x double> %1
  ret <vscale x 2 x double> %2
}

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (ssijaric-nv) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2881]: Thanks for having a look, Yingwei.  I'll test and upload the revised version.

Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2881]: The IRBuilder should do this like in other cases?

**Comment by nikic** [comment]
> > > 1. if we are dealing with scalable vector constants as there is no
> > >    immediate representation for a negative scalable zeroinitializer.
> > 
> > 
> > I thought there was a splat constant for these, this would just be a splat of -0
> 
> Thanks for having a look, Matt. Would a better approach be to generate the following, instead of giving up?
> 
> define <vscale x 2 x double> @test_fneg_select_svec(<vscale x 2 x i1> %cond, <vscale x 2 x double> %b) { %1 = fneg fast <vscale x 2 x double> %b %2 = select <vscale x 2 x i1> %cond, <vscale x 2 x double> splat(double -0.0), <vscale x 2 x double> %1 ret <vscale x 2 x double> %2 }

Yes. It looks like ConstantFoldUnaryInstruction is currently missing splat support for scalable vectors. Adding that should resolve the issue.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2887]: Right, just the changes to ConstantFold should be sufficient, after that there should not be a need to touch this InstCombine code anymore...

**Review by nikic — APPROVED** [review]
LGTM

**Comment by ssijaric-nv** [comment]
Thank you, all.  Can someone please merge this change, as I don't have write access.  I'll rebase and push again, to get a clean build before merging.

I ran the test-suite locally with -O3 on AArch64 and x86 machines:

AArch64:
Total Discovered Tests: 3425
  Passed: 3425 (100.00%)

With -mcpu=neoverse-v2
Total Discovered Tests: 3425
  Passed: 3425 (100.00%)

x86:
Total Discovered Tests: 2441
  Passed: 2441 (100.00%)

ninja check-llvm passed as well.

**Pull request merged**
